src/textinput/TextInput.ts

```diff
@@ -0,0 +1,161 @@
+	minLength?: number | string;
+	name?: string;
+	placeholder?: string;
+	readOnly?: boolean;
+	required?: boolean;
+	type?: TextInputType;
+	value?: string;
+	onBlur?(event: FocusEvent): void;
+	onChange?(event: Event): void;
+	onClick?(event: MouseEvent): void;
+	onFocus?(event: FocusEvent): void;
+	onInput?(event: Event): void;
+	onKeyDown?(event: KeyboardEvent): void;
+	onKeyPress?(event: KeyboardEvent): void;
+	onKeyUp?(event: KeyboardEvent): void;
+	onMouseDown?(event: MouseEvent): void;
+	onMouseUp?(event: MouseEvent): void;
+	onTouchStart?(event: TouchEvent): void;
+	onTouchEnd?(event: TouchEvent): void;
+	onTouchCancel?(event: TouchEvent): void;
+}
+
+export const TextInputBase = ThemeableMixin(WidgetBase);
+
+@theme(css)
+export default class TextInput extends TextInputBase<TextInputProperties> {
+	onBlur (event: FocusEvent) { this.properties.onBlur && this.properties.onBlur(event); }
+	onChange (event: Event) { this.properties.onChange && this.properties.onChange(event); }
+	onClick (event: MouseEvent) { this.properties.onClick && this.properties.onClick(event); }
+	onFocus (event: FocusEvent) { this.properties.onFocus && this.properties.onFocus(event); }
+	onInput (event: Event) { this.properties.onInput && this.properties.onInput(event); }
+	onKeyDown (event: KeyboardEvent) { this.properties.onKeyDown && this.properties.onKeyDown(event); }
+	onKeyPress (event: KeyboardEvent) { this.properties.onKeyPress && this.properties.onKeyPress(event); }
+	onKeyUp (event: KeyboardEvent) { this.properties.onKeyUp && this.properties.onKeyUp(event); }
+	onMouseDown (event: MouseEvent) { this.properties.onMouseDown && this.properties.onMouseDown(event); }
+	onMouseUp (event: MouseEvent) { this.properties.onMouseUp && this.properties.onMouseUp(event); }
+	onTouchStart (event: TouchEvent) { this.properties.onTouchStart && this.properties.onTouchStart(event); }
+	onTouchEnd (event: TouchEvent) { this.properties.onTouchEnd && this.properties.onTouchEnd(event); }
+	onTouchCancel (event: TouchEvent) { this.properties.onTouchCancel && this.properties.onTouchCancel(event); }
+
+	render() {
+		const {
+			describedBy,
+			disabled,
+			formId,
+			invalid,
+			label,
+			maxLength,
+			minLength,
+			name,
+			placeholder,
+			readOnly,
+			required,
+			type = 'text',
+			value
+		} = this.properties;
+
+		const stateClasses = [
+			disabled ? css.disabled : null,
+			typeof invalid === 'boolean' ? invalid ? css.invalid : css.valid : null,
+			readOnly ? css.readonly : null,
+			required ? css.required : null
+		];
+
+		const textinput = v('div', { classes: this.classes(css.inputWrapper) }, [
+			v('input', {
+				bind: this,
+				classes: this.classes(css.root).get(),
+				'aria-describedby': describedBy,
+				disabled,
+				'aria-invalid': invalid,
+				maxlength: maxLength ? maxLength + '' : null,
+				minlength: minLength ? minLength + '' : null,
+				name,
+				placeholder,
+				readOnly,
+				'aria-readonly': readOnly ? 'true' : null,
+				required,
```

Comment:
Yup, it has an extra border, but I should add in the label `*` too. Forgot about that :)